### PR TITLE
header mainmanager added

### PR DESCRIPTION
**Pull request by younger15**

### CODE REVIEW

**Comment by younger15** [comment]
code style changed

File: src/Header/MainManager.h [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+// found in the LICENSE file.
+
+#ifndef Group4_Header_MainManager_H_
+#define Group4_Header_MainManager_H_
+
+#include "HttpListener.h"
+#include "RequestHandler.h"
+#include "SocketProcess.h"
+
+#include<vector>
```
Comment (john0312) [inline src/Header/MainManager.h]: space after include for consistency, i.e.
#include <vector>

File: src/Header/MainManager.h [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+//will initialize http listener, do multiprocess when listener callback
+class MainManager
+{
+public:
+    //httpListener callback, do fork and new requestHandler to call api, child process add to parents vector for managing.
```
Comment (john0312) [inline src/Header/MainManager.h:20]: Do you want to delete the copy constructor/assignment operator?

File: src/Header/MainManager.h (hunk 1 above)
Comment (younger15) [reply src/Header/MainManager.h]: Done

File: src/Header/MainManager.h (hunk 2 above)
Comment (younger15) [reply src/Header/MainManager.h:20]: Done

File: src/Header/MainManager.h [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    void NewRequest(const string &requestStr);
+    //requestHandler callback, remove process from parents vector.
+    void EndRequest();
+    //for managing all processes.
+    std::unordered_map<struct sockaddr_in, pid_t> socketMap;
```
Comment (john0312) [inline src/Header/MainManager.h]: Does this compile?

File: src/Header/MainManager.h [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    //singleton code convention method, listen port number is required.
+    static MainManager* GetInstance(const int &portNum);
+
+    //Delete copy constructor
+    MainManager(const MainManager &cpy) = delete;
+    //Delete assignment operator
+    MainManager& operator = (const MainManager &cpy) = delete;
```
Comment (john0312) [inline src/Header/MainManager.h]: Please remove extra space. `operator=(...`

File: src/Header/MainManager.h [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+private:
+    //singleton code convention method.
+    MainManager();
+    //used to record which port is listening, one process will listen one port
+    //will call httpListener.BindPort()
+    int portNum;
```
Comment (john0312) [inline src/Header/MainManager.h]: Please use underscore for naming. `port_num`
https://google.github.io/styleguide/cppguide.html#Variable_Names

File: src/Header/MainManager.h [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+{
+public:
+    //httpListener callback, do fork and new requestHandler to call api, child process add to parents vector for managing.
+    void NewRequest(const string &requestStr);
+    //requestHandler callback, remove process from parents vector.
+    void EndRequest();
+    //for managing all processes.
+    std::unordered_map<struct sockaddr_in, pid_t> socketMap;
+    //singleton code convention method, listen port number is required.
+    static MainManager* GetInstance(const int &portNum);
```
Comment (john0312) [inline src/Header/MainManager.h]: Please add a new InitInstance() for singleton that requires initialization.
You can assert() if GetInstance() is called without InitInstance.

Comment (john0312) [reply src/Header/MainManager.h]: Also, use uint16_t for port number.

File: src/Header/MainManager.h (hunk 4 above)
Comment (younger15) [reply src/Header/MainManager.h]: Done

File: src/Header/MainManager.h (hunk 6 above)
Comment (younger15) [reply src/Header/MainManager.h]: Done

File: src/Header/MainManager.h (hunk 5 above)
Comment (younger15) [reply src/Header/MainManager.h]: Done

File: src/Header/MainManager.h (hunk 3 above)
Comment (younger15) [reply src/Header/MainManager.h]: changed map to <int(socket fd), pid_t(process id)>

File: src/Header/MainManager.h [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+#include <netinet/in.h>
+
+#include <vector>
+#include <unordered_map>
+
+#include "HttpListener.h"
+#include "RequestHandler.h"
+
+
```
Comment (john0312) [inline src/Header/MainManager.h]: Please remove extra blank lines here. You can use clang-format to do it.

File: src/Header/MainManager.h [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+//will initialize http listener, do multiprocess when listener callback
+class MainManager
+{
+public:
+    //httpListener callback, do fork and new requestHandler to call api, child process add to parents vector for managing.
+    void NewRequest(const string &requestStr);
```
Comment (john0312) [inline src/Header/MainManager.h]: Please use underscore for naming: `request_str`
https://google.github.io/styleguide/cppguide.html#Variable_Names

File: src/Header/MainManager.h [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    //requestHandler callback, remove process from parents vector.
+    void EndRequest();
+    //for managing all processes.
+    //socket file descripter => pid_t
+    std::unordered_map<int, pid_t> socketMap;
```
Comment (john0312) [inline src/Header/MainManager.h]: Same here. `socket_map`

File: src/Header/MainManager.h [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+    //Initlize singleton instance
+    static void InitInstance(const uint16_t &portNum);
```
Comment (john0312) [inline src/Header/MainManager.h]: `port_num`

File: src/Header/MainManager.h [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+    //Delete copy constructor
+    MainManager(const MainManager &cpy) = delete;
+    //Delete assignment operator
+    MainManager& operator=(const MainManager &cpy) = delete;
+private:
+    //singleton code convention method.
+    MainManager();
+    //used to record which port is listening, one process will listen one port
+    //will call httpListener.BindPort()
+    uint16_t port_Num;
```
Comment (john0312) [inline src/Header/MainManager.h]: `port_num`

**Review by john0312 — COMMENTED** [review]
Please clang-format your PR.

File: src/Header/MainManager.h (hunk 7 above)
Comment (younger15) [reply src/Header/MainManager.h]: Done

File: src/Header/MainManager.h (hunk 8 above)
Comment (younger15) [reply src/Header/MainManager.h]: Done

File: src/Header/MainManager.h (hunk 10 above)
Comment (younger15) [reply src/Header/MainManager.h]: Done

File: src/Header/MainManager.h (hunk 9 above)
Comment (younger15) [reply src/Header/MainManager.h]: Done

File: src/Header/MainManager.h (hunk 11 above)
Comment (younger15) [reply src/Header/MainManager.h]: Done

File: src/Header/MainManager.h [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+#include <netinet/in.h>
+
+#include <vector>
+#include <unordered_map>
+
+#include "HttpListener.h"
+#include "RequestHandler.h"
+
+//will initialize http listener, do multiprocess when listener callback
+class MainManager
```
Comment (john0312) [inline src/Header/MainManager.h:17]: Please comment that the child process will not use this class. You can do it in a separate PR.

**Review by john0312 — APPROVED** [review]
